### PR TITLE
Feature/api code generation as gradletask

### LINKED ISSUE #86: Add API code generation as gradle task
**Describe the problem or limitation you are having in your project:**

For the merge purpose, we decided to set api code generation as a gradle task.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**

Not appropriate

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

See @raniejade repository

**Describe implementation detail for your proposal (in code), if possible:**

Move code to `buildSrc` of project.
Add a task implementation.
Call task from `godot-library` gradle.

**If this enhancement will not be used often, can it be worked around with a few lines of code?:**

Not appropriate.

**Is there a reason why this should be in this project and not individually solved?:**

Not appropriate.

### PR DESCRIPTION
**Pull request by piiertho**
This implements new api generation code.
It uses:
- new core types
- new signals
- new Variant wrapping.

API generation is now a gradle task. godot-library:build gradle task is dependant of the api generation task.
Singletons are now generated as kotlin `object`.
Classes not instantiable inheriting from a Singleton are not generated anymore.

This resolves #86

### CODE REVIEW

File: godot-kotlin/godot-library/build.gradle.kts [hunk 1]
```diff
@@ -49,6 +50,16 @@ kotlin {
     }
 }
 
+val generateAPI by tasks.creating(GenerateApiTask::class) {
+    source.set(project.file("$rootDir/godot-kotlin/godot-headers/api.json"))
+    outputDirectory.set(project.file("$rootDir/godot-kotlin/godot-library/src/nativeGen/kotlin/"))
+}
+
+tasks {
```
Comment (raniejade) [inline godot-kotlin/godot-library/build.gradle.kts]: A more type-safe approach is to use `tasks.withType<KotlinCompile> { ... }`

File: godot-kotlin/godot-library/src/nativeGen/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+kotlin/*
```
Comment (raniejade) [inline godot-kotlin/godot-library/src/nativeGen/.gitignore:1]: this is to ignore the generated sources right?

File: godot-kotlin/godot-library/src/nativeInternal/kotlin/godot/internal/utils/getFromGDNative.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+internal fun getConstructor(className: String, oldClassName: String): COpaquePointer {
+    memScoped {
+        return Godot.gdnative.godot_get_class_constructor!!.invoke(oldClassName.cstr.ptr) ?: throw NotImplementedError("Cannot get constructor for class $className")
+    }
+}
+
+internal fun getSingleton(className: String, oldClassName: String): COpaquePointer {
+    memScoped {
+        return Godot.gdnative.godot_global_get_singleton!!.invoke(oldClassName.cstr.ptr) ?: throw NullPointerException("Cannot get singleton instance $className")
```
Comment (raniejade) [inline godot-kotlin/godot-library/src/nativeInternal/kotlin/godot/internal/utils/getFromGDNative.kt]: `NotImplementedError`? You should never throw a NPE.

**Review by raniejade — CHANGES_REQUESTED** [review]

File: godot-kotlin/godot-library/src/nativeGen/.gitignore (hunk 2 above)
Comment (piiertho) [reply godot-kotlin/godot-library/src/nativeGen/.gitignore:1]: yep

File: godot-kotlin/godot-library/src/nativeInternal/kotlin/godot/internal/utils/getFromGDNative.kt (hunk 3 above)
Comment (piiertho) [reply godot-kotlin/godot-library/src/nativeInternal/kotlin/godot/internal/utils/getFromGDNative.kt]: yep you're right, seems to be a mistake, I wanted NotImplemented as others in file ...

File: buildSrc/src/main/kotlin/godot/codegen/TypeCast.kt [hunk 4]
```diff
@@ -0,0 +1,162 @@
+package godot.codegen
```
Comment (chippmann) [inline buildSrc/src/main/kotlin/godot/codegen/TypeCast.kt:1]: I would cleanup/resolve some of the todo's in this file.

File: gradle.properties [hunk 5]
```diff
@@ -1,4 +1,4 @@
 #kotlin.mpp.enableGranularSourceSetsMetadata=true
 # kotlin native compiler is now integrated with the gradle daemon, however, the default memory
 # settings for the daemon is not enough for kotlin native. This setting bumps the maximum memory to 3gb.
-org.gradle.jvmArgs=-Xmx3G
\ No newline at end of file
+org.gradle.jvmargs=-Xmx3G
```
Comment (chippmann) [inline gradle.properties:4]: Are you sure about that?

**Review by chippmann — CHANGES_REQUESTED** [review]
Really awesome work. But one problem (i guess)
There are classes generated which don't belong to godot (as far as i can see). At least i don't see them in the api json:
suspiciousClasses

These are just some examples of what i mean.

EDIT: It seems they are coming from the `api.json` but curiously they are not present in the `gdnative_api.json` or in the `gdnative_api_struct.gen.h`

EDIT2: Short summary of discussion on Discord: It seems these classes are for android plugin development pre 3.x.
@piiertho I suggest we make some sort of Blacklist to exclude such classes from being generated as they are not present in gdnative anyways. What do you think?

File: gradle.properties (hunk 5 above)
Comment (chippmann) [reply gradle.properties:4]: just saw that you are right

File: godot-kotlin/godot-library/build.gradle.kts (hunk 1 above)
Comment (piiertho) [reply godot-kotlin/godot-library/build.gradle.kts]: Done ! :)

File: godot-kotlin/godot-library/src/nativeInternal/kotlin/godot/internal/utils/getFromGDNative.kt (hunk 3 above)
Comment (piiertho) [reply godot-kotlin/godot-library/src/nativeInternal/kotlin/godot/internal/utils/getFromGDNative.kt]: Done !

File: buildSrc/src/main/kotlin/godot/codegen/TypeCast.kt (hunk 4 above)
Comment (piiertho) [reply buildSrc/src/main/kotlin/godot/codegen/TypeCast.kt:1]: Done !

**Review by chippmann — APPROVED** [review]

**Review by CedNaru — APPROVED** [review]

**Pull request merged**